[AIRFLOW-464] Add setdefault method to Variable

btallman opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- _https://issues.apache.org/jira/browse/AIRFLOW-464_

This was changed from adding a create_if_none flag to Variable.get based on feedback on this PR.

Testing Done:
- Added a test to test/core.py to cover this functionality

gwax commented on airflow/models.py:
```diff
@@ -3367,6 +3367,36 @@ def val(cls):
+        obj = session.query(cls).filter(cls.key == key).first()
+        if obj is None:
+            if default is not None:
+                Variable.set(key, default, serialize_json=deserialize_json)
```
do we need to commit the session?

gwax commented on airflow/models.py:
```diff
@@ -3367,6 +3367,36 @@ def val(cls):
+         and un-encode it when retrieving a value
+        :return: Mixed
+        """
+        obj = session.query(cls).filter(cls.key == key).first()
```
Would it make sense to reuse the `get` logic directly. Something like:

``` python
exists = session.query(literal(True)).where(
    session.query(cls).filter(cls.key == key).exists()
    ).scalar()
if not exists:
    Variable.set(key, default, serialize_json=deserialize_json)
return Variable.get(key, deserialize_json=deserialize_json)
```

(I haven't tested the exist query)

btallman replied:
Well, that would incur an extra DB query... Another option, which doesn't require an extra round trip and has no try/except overhead is like this...

```
    @classmethod
    def setdefault(cls, key, default, deserialize_json=False):
        """
        Like a Python builtin dict object, setdefault returns the current value
        for a key, and if it isn't there, stores the default value and returns it.

        :param key: Dict key for this Variable
        :type key: String
        :param: default: Default value to set and return if the variable
        isn't already in the DB
        :type: default: Mixed
        :param: deserialize_json: Store this as a JSON encoded value in the DB
         and un-encode it when retrieving a value
        :return: Mixed
        """
        false_obj = {}
        obj = Variable.get(key, default_var=false_obj, deserialize_json=False)
        if obj == false_obj:
            if default is not None:
                Variable.set(key, default, serialize_json=deserialize_json)
                return default
            else:
                raise ValueError('Default Value must be set')
        else:
            if deserialize_json:
                return json.loads(obj.val)
            else:
                return obj.val

```

This works because with deserialize_json=False, you will never return an object... So, no new query logic...

gwax replied:
In that case it would be safer to use a proper sentinel instead of an empty dict:

``` python
default_sentinel = object()
obj = Variable.get(key, default_var=default_sentinel)  # deserialize_json=False is the default, right?
if obj is default_sentinel:
    ...
```

Though, at this point, maybe it's better to ask forgiveness rather than permission:

``` python
try:
    return Variable.get(key, deserialize_json=deserialize_json)
except ValueError:
    Variable.set(key, default_var, serialize_json=deserialize_json)
    return default_var
```

(as an aside, looking at this says to me that `Variable.get` should be raising a `KeyError` instead of a `ValueError`)

btallman replied:
This seems clean, so pushed improved code

btallman replied:
I actually changed this again to completely use Variable.get and Variable.set.

Basically, I pass in an Object as the default with `deserialize_json=False`, and if I get the Object back, I KNOW it didn't find a value, in which case I set it and return it, otherwise, I decode the value, if the `deserialize_json=True`, and return it:

```
    @classmethod
    def setdefault(cls, key, default, deserialize_json=False):
        """
        Like a Python builtin dict object, setdefault returns the current value
        for a key, and if it isn't there, stores the default value and returns it.

        :param key: Dict key for this Variable
        :type key: String
        :param: default: Default value to set and return if the variable
        isn't already in the DB
        :type: default: Mixed
        :param: deserialize_json: Store this as a JSON encoded value in the DB
         and un-encode it when retrieving a value
        :return: Mixed
        """
        false_obj = {}
        obj = Variable.get(key, default_var=false_obj, deserialize_json=False)
        if obj == false_obj:
            if default is not None:
                Variable.set(key, default, serialize_json=deserialize_json)
                return default
            else:
                raise ValueError('Default Value must be set')
        else:
            if deserialize_json:
                return json.loads(obj.val)
            else:
                return obj.val

```

btallman committed:
Add setdefault method to Variable

In order to assist with environment migrations, we added a setdefault method to
the Variable object. This allows default variables to be created (and then
edited) with less chance for typos/copy+paste bugs.

Also changed Variable.get ValueError exception to a KeyError

Variable.setdefault(key, default, deserialize_json=[True|False]) returns either the
value stored in Variable(key) or sets Variable(key) = default and returns default.
